README.md
# music.note.height

[![Build Status](https://travis-ci.org/danigb/music.note.height.svg?branch=master)](https://travis-ci.org/danigb/music.note.height)
[![Test Coverage](https://codeclimate.com/github/danigb/music.note.height/badges/coverage.svg)](https://codeclimate.com/github/danigb/music.note.height/coverage)
[![Coverage](https://codeclimate.com/github/danigb/music.note.height/badges/gpa.svg)](https://codeclimate.com/github/danigb/music.note.height)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)
[![npm version](https://img.shields.io/npm/v/music.note.height.svg)](https://www.npmjs.com/package/music.note.height)
[![license](https://img.shields.io/npm/l/music.note.height.svg)](https://www.npmjs.com/package/music.note.height)
[![distribution](https://img.shields.io/badge/dist-2.3kb-blue.svg)](https://github.com/danigb/music.kit)
[![music.kit](https://img.shields.io/badge/music-kit-yellow.svg)](https://github.com/danigb/music.kit)

`music.note.height` is a small (2.3kb) and fast library to convert between midi numbers <-> note names <-> frequencies. It's the bridge between midi, your app and your synthetizers.

```js
var note = require('music.note.height')
// get a note from midi
var name = note.fromMidi(69) // => 'A4'
// write it to the console
console.log(name) // => prints 'A4'
// give it to your synth
note.toFreq(name) // => 440
```

This is part of [music.kit](https://github.com/danigb/music.kit)

## Features

- Note name to midi and midi to note name conversion
- Note name to frequency and frequency to note name conversion
- Frequency conversion with custom tunnings (different from A4 = 440hz)
- Use scientific pitch by default. You can use your own note name parser
- Calculate frequency deviations in cents

## Installation

Install via npm: `npm install --save music.note.height` and require it. Or grab the browser ready distribution file [here](https://github.com/danigb/music.note.height/blob/master/dist/music.note.height.min.js)

## Usage

#### Note names

The `note` function returns the [scientific notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation) of a given note if valid. Can be used to check if its a valid note:

```js
note('bbb') // => 'Bbb'
note('fx4') // => 'F##4'

if (note(str) !== null) { /* valid pitch str */ }
```

#### Working with midi

You have two functions for converting from and to midi numbers:

```js
note.toMidi('A4') // => '69'
note.fromMidi(69) // => 'A4'
```

#### Working with frequencies

The same way, you have two frequency related functions:

```js
note.toFreq('A4') // => 440
note.fromFreq(440) // => 'A4'
```

#### Using different pitch notation

In the case scientific notation is not what you need, you can always use [pitch array notation](https://github.com/danigb/pitch-array) for every function that expects a string:

```js
note.toFreq([5, 0, 3]) // => 220
```

Also you can covert from scientific notation to pitch array notation with the `parse` function:

```js
note.parse('A3') // => [5, 0, 3]
```

#### More...

This is part of [music.kit](https://github.com/danigb/music.kit)

## Documentation

[Generated documentation here](https://github.com/danigb/music.note.height/blob/master/API.md)

## License

MIT License


CHANGELOG
CHANGELOG

[0.0.4] Reduce dependencies to music.note.parser
[0.0.3] Custom tuning version of functions (toFreq is now mapeable)
[0.0.2] First version

DIST
[0.0.4] 2878
[0.0.3] 4665
[0.0.2] 4335

BENCHMARKS


test/note-test.js
var vows = require('vows')
var assert = require('assert')
var note = require('../')

vows.describe('music.note.height').addBatch({
  'parse': function () {
    assert.deepEqual(note.parse('C#4'), [0, 1, 4, 0])
    assert.deepEqual(note.parse('Bb'), [6, -1])
  },
  'note': function () {
    assert.equal(note('c'), 'C')
    assert.equal(note('bbb3'), 'Bbb3')
    assert.equal(note('fx'), 'F##')
    assert.equal(note([0, 1, 3, 0]), 'C#3')
  },
  'midi': {
    'fromMidi': function () {
      var notes = [60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73]
      .map(note.fromMidi).join(' ')
      assert.equal(notes, 'C4 Db4 D4 Eb4 E4 F4 F#4 G4 Ab4 A4 Bb4 B4 C5 Db5')
    },
    'toMidi': function () {
      assert.deepEqual('C4 D4 E4 F4 G4 A4 B4 C4'.split(' ').map(note.toMidi), [60, 62, 64, 65, 67, 69, 71, 60])
      assert.deepEqual('C#4 D#4 E#4 F#4 G#4 A#4 B#4 C#4'.split(' ').map(note.toMidi), [61, 63, 65, 66, 68, 70, 72, 61])
      assert.deepEqual('C##4 D##4 E##4 F##4 G##4 A##4 B##4 C##4'.split(' ').map(note.toMidi), [62, 64, 66, 67, 69, 71, 73, 62])
      assert.deepEqual('Cb4 Db4 Eb4 Fb4 Gb4 Ab4 Bb4 Cb4'.split(' ').map(note.toMidi), [59, 61, 63, 64, 66, 68, 70, 59])
      assert.deepEqual('Cbb3 Dbb3 Ebb3 Fbb3 Gbb3 Abb3 Bbb3 Cbb3'.split(' ').map(note.toMidi), [46, 48, 50, 51, 53, 55, 57, 46])
    },
    'toMidi: note class does not have midi': function () {
      assert.deepEqual('C D E F G A B C'.split(' ').map(note.toMidi), [ null, null, null, null, null, null, null, null ])
    },
    'toMidi: enharmonics': function () {
      assert.equal(note.toMidi('B#3'), note.toMidi('C4'))
      assert.equal(note.toMidi('B##3'), note.toMidi('Db4'))
    }
  }
}).export(module)


test/frequency-test.js
var vows = require('vows')
var assert = require('assert')
var note = require('../')

vows.describe('frequencies').addBatch({
  'toFreq': {
    'toFreq': function () {
      assert.equal(note.toFreq('A4'), 440)
      assert.equal(note.toFreq('A3'), 220)
      assert.equal(note.toFreq('E4'), 329.6275569128699)
      assert.equal(note.toFreq('F4'), 349.2282314330039)
    },
    'map notes': function () {
      assert.deepEqual(['C3', 'D3', 'e3'].map(note.toFreq),
        [ 130.8127826502993, 146.8323839587038, 164.81377845643496 ])
    }
  },
  'fromFreq': function () {
    assert.equal(note.fromFreq(440), 'A4')
    assert.equal(note.fromFreq(220), 'A3')
    assert.equal(note.fromFreq(329.6275569128699), 'E4')
    assert.equal(note.fromFreq(330), 'E4')
    assert.equal(note.fromFreq(335), 'E4')
    assert.equal(note.fromFreq(340), 'F4')
    assert.equal(note.fromFreq(349.2282314330039), 'F4')
  },
  'fromCustomFreq with custom tuning': function () {
    assert.equal(note.fromCustomFreq(220, 110), 'A3')
  },
  'toCustomFreq partially applied': function () {
    assert.equal(note.toCustomFreq(440)('a3'), 220)
    assert.equal(note.toCustomFreq(444)('a3'), 222)
  },
  'toFreq custom tuning': function () {
    assert.equal(note.toCustomFreq(444, 'A4'), 444)
    assert.equal(note.toCustomFreq(222, 'A3'), 111)
  },
  'cents': function () {
    assert.equal(note.cents('A4', 'A#4'), 100)
    assert.equal(note.cents('A4', 444), 15.66)
  }
}).export(module)


index.js
'use strict'

var parser = require('music.note.parser')

// Semitones from C to C D E F G A B
var SEMITONES = [ 0, 2, 4, 5, 7, 9, 11 ]
// Chromatic melodic scale
var CHROMATIC = [ 'C', 'Db', 'D', 'Eb', 'E', 'F', 'F#', 'G', 'Ab', 'A', 'Bb', 'B' ]

function arr (p) { return Array.isArray(p) ? p : parser(p) }

/**
 * Get the note in scienific notation or null if not a valid note
 *
 * @param {String} note - the note
 * @return {String} the scientific notation string or null if not a valid note
 *
 * @example
 * note('c#') // => 'C#'
 * note('fx') // => 'F##'
 * note('bbb') // => 'Bbb'
 * note('m') // => null
 */
function note (src) {
  return parser.build(arr(src))
}

note.parse = function (note) {
  var p = arr(note)
  return p && p.length !== 3 ? p : null
}
note.build = parser.build

/**
 * Get the pitch of the given midi number
 *
 * This method doesn't take into account diatonic spelling. Always the same
 * pitch class is given to the same midi number.
 *
 * @name fromMidi
 * @function
 * @param {Integer} midi - the midi number
 * @return {String} the pitch
 *
 * @example
 * note.fromMidi(69) // => 'A4'
 */
note.fromMidi = function (midi) {
  var name = CHROMATIC[midi % 12]
  var oct = Math.floor(midi / 12) - 1
  return name + oct
}

/**
 * Get the midi number of a pitch
 *
 * @name toMidi
 * @function
 * @param {String|Array} pitch - the pitch string (or pitch array)
 * @return {Integer} the midi number
 *
 * @example
 * note.toMidi('A4') // => 69
 * note.toMidi('A3') // => 57
 */
note.toMidi = function (note) {
  var p = Array.isArray(note) ? note : parser(note)
  if (!p || (!p[2] && p[2] !== 0)) return null
  return SEMITONES[p[0]] + p[1] + 12 * (p[2] + 1)
}

/**
 * Get the pitch of a given frequency using a custom tuning
 *
 * It will round the frequency to the nearest pitch frequency. Use `cents` function
 * if you need to know the difference between the the frequency and the note.
 *
 * This function can be partially applied
 *
 * @name fromCustomFreq
 * @function
 * @param {Float} tuning - the frequency of A4
 * @param {Float} freq - the frequency of the note you want
 * @return {String} the note name
 *
 * @see fromFreq
 * @see cents
 *
 * @example
 * var fromFreq = note.fromCustomFreq(444)
 * fromFreq(222) // => 'A3'
 */
note.fromCustomFreq = function (t, f) {
  var tuning = t || 440
  if (arguments.length > 1) return note.fromCustomFreq(tuning)(f)

  return function (freq) {
    var lineal = 12 * ((Math.log(freq) - Math.log(tuning)) / Math.log(2))
    var midi = Math.round(69 + lineal)
    return note.fromMidi(midi)
  }
}

/**
 * Get the pitch of a given frequency.
 *
 * It will round the frequency to the nearest pitch frequency. Use `cents` function
 * if you need to know the difference between the the frequency and the note.
 *
 * @name fromFreq
 * @function
 * @param {Float} freq - the frequency
 * @return {String} the pitch
 *
 * @see fromCustomFreq
 *
 * @example
 * note.fromFreq(440) // => 'A4'
 * note.fromFreq(443) // => 'A4'
 * note.cents(443, 'A4') // => ... to get the difference
 */
note.fromFreq = note.fromCustomFreq(440)

// decimal number
var NUM = /^\d+(?:\.\d+)?$/
/**
 * Get the pitch frequency in herzs with custom concert tuning
 *
 * This function is currified so it can be partially applied
 *
 * @param {Float} tuning - the frequency of A4 (null means 440)
 * @param {String|Array} pitch - the note name
 * @return {Float} the frequency of the note
 *
 * @example
 * var toFreq = note.toCustomFreq(444)
 * toFreq('A2') // => 111
 */
note.toCustomFreq = function (tuning, p) {
  tuning = tuning || 440
  if (arguments.length > 1) return note.toCustomFreq(tuning)(p)
  return function (p) {
    if (NUM.test(p)) return +p
    var midi = note.toMidi(p)
    if (!midi) return null
    return Math.pow(2, (midi - 69) / 12) * tuning
  }
}

/**
 * Get the pitch frequency in hertzs (using 440 as concert pitch)
 *
 * @name toFreq
 * @function
 * @param {String} pitch - the pitch
 * @return {Float} - the pitch frequency
 *
 * @example
 * note.toFreq('A4') // => 440
 * note.toFreq('A3') // => 222
 */
note.toFreq = note.toCustomFreq(440)

/**
 * Get the distance in cents between pitches or frequencies
 *
 * @name cents
 * @function
 * @param {String|Integer} from - first pitch or frequency
 * @param {String|Integer} to - other pitch or frequency
 * @param {Integer} decimals - the decimal precision (2 by default)
 * @return {Integer} the distance in cents
 *
 * @example
 * note.cents(440, 444) // => 15.66
 * note.cents('A4', 444) // => 15.66
 * note.cents('A4', 'A#4') // => 100
 */
note.cents = function (from, to, decimals) {
  var dec = decimals ? Math.pow(10, decimals) : 100
  var fromFreq = note.toFreq(from)
  var toFreq = note.toFreq(to)
  return Math.floor(1200 * (Math.log(toFreq / fromFreq) * dec / Math.log(2))) / dec
}

if (typeof module === 'object' && module.exports) module.exports = note
if (typeof window !== 'undefined') window.note = note
